File: Cadena de Hoteles/Hotel/src/main/java/org/example/controllers/PagoController.java
package org.example.controllers;

import io.javalin.Javalin;
import io.javalin.http.Context;
import org.example.dtos.PagoRequestDTO;
import org.example.services.PagoService;

import java.util.Map;

/**
 * Controller que gestiona el procesamiento de pagos de reservaciones para usuarios autenticados.
 * Requiere sesion activa; el usuarioId se obtiene del contexto inyectado por el middleware JWT.
 */
public class PagoController {

    private final PagoService pagoService;

    /**
     * Crea una instancia de PagoController con sus dependencias inyectadas.
     */
    public PagoController(PagoService pagoService) {
        this.pagoService = pagoService;
    }

    /**
     * Registra la ruta de pago en la aplicacion Javalin.
     * @param app instancia de Javalin donde se registra la ruta.
     */
    public void registerRoutes(Javalin app) {

        // Procesa el pago de una reservacion perteneciente al usuario autenticado
        app.post("/reservaciones/{id}/pago", this::handleProcesarPago);
    }

    void handleProcesarPago(Context ctx) {
        int usuarioId     = ctx.attribute("usuarioId");
        int reservacionId = Integer.parseInt(ctx.pathParam("id"));
        PagoRequestDTO request = ctx.bodyAsClass(PagoRequestDTO.class);

        try {
            var resultado = pagoService.procesarPago(reservacionId, usuarioId, request,
                    ctx.ip(), ctx.userAgent());
            ctx.status(200).json(resultado);
        } catch (IllegalArgumentException e) {
            ctx.status(400).json(Map.of("mensaje", e.getMessage()));
        } catch (RuntimeException e) {
            ctx.status(500).json(Map.of("mensaje", e.getMessage()));
        }
    }
}

File: Agencia de Viajes/API Agencia/agencia-viajes/internal/controllers/PagoController.go
// # Package controllers
//
// Controladores HTTP de la API de Movent. Cada controlador agrupa los handlers
// relacionados a un recurso o dominio especifico de la aplicacion.
package controllers

import (
	"agencia-viajes/internal/dto"
	"agencia-viajes/internal/helpers"
	"agencia-viajes/internal/services"
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
)

// PagoController
//
// Controlador que maneja los endpoints relacionados al procesamiento de pagos
// de reservaciones existentes en la plataforma.
type PagoController struct {
	service   *services.PagoService
	logSesion *services.LogSesionService
}

// NewPagoController
//
// Constructor que retorna una nueva instancia de PagoController
// con el servicio de pagos y el servicio de log inyectados.
//
// Parametros:
//   - service: puntero al servicio de pagos
//   - logSesion: puntero al servicio de log de sesion para auditoria
//
// Retorna:
//   - *PagoController: puntero a la nueva instancia
func NewPagoController(service *services.PagoService, logSesion *services.LogSesionService) *PagoController {
	return &PagoController{service: service, logSesion: logSesion}
}

// Pagar
//
// Procesa el pago de una reservacion existente para el usuario autenticado.
// Lee el ID del usuario desde el contexto de Gin verificando su existencia,
// valida el body del request y delega el procesamiento al servicio de pagos.
// Registra en log_sesion el resultado del intento de pago (exitoso o fallido).
//
// Parametros:
//   - c: contexto de Gin con la solicitud HTTP
//
// Retorna:
//   - HTTP 200 OK: JSON con mensaje de confirmacion si el pago fue exitoso
//   - HTTP 400 Bad Request: si los datos de pago en el body estan incompletos o son invalidos
//   - HTTP 401 Unauthorized: si el usuario no esta autenticado (usuario_id ausente en contexto)
//   - HTTP 422 Unprocessable Entity: si el servicio rechaza el pago (fondos insuficientes,
//     reservacion no encontrada, etc.)
//
// Notas:
//   - ID 25 (COMPRA_EXITOSA) se registra con el no_reservacion como loginIntentado
//   - ID 26 (COMPRA_FALLIDA_PAGO) se registra con "reserva_id=N" como loginIntentado
func (ctrl *PagoController) Pagar(c *gin.Context) {
	var req dto.PagoReservacionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"mensaje": "Datos inválidos"})
		return
	}

	// FIX: verificar existencia del usuario_id (antes causaba panic con type assertion sin check)
	usuarioIDRaw, exists := c.Get("usuario_id")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"mensaje": "No autenticado"})
		return
	}
	usuarioID := usuarioIDRaw.(int)

	noReservacion, err := ctrl.service.ProcesarPago(usuarioID, req)
	if err != nil {
		uid := usuarioID
		loginInt := fmt.Sprintf("reserva_id=%d", req.ReservacionID)
		ctrl.logSesion.Registrar(c, helpers.TipoCompraFallidaPago, &uid, loginInt, err.Error())
		c.JSON(http.StatusUnprocessableEntity, gin.H{"mensaje": err.Error()})
		return
	}

	uid := usuarioID
	ctrl.logSesion.Registrar(c, helpers.TipoCompraExitosa,
		&uid, noReservacion,
		fmt.Sprintf("Compra procesada para reserva %s", noReservacion))

	c.JSON(http.StatusOK, gin.H{
		"mensaje": "Pago procesado y reservación confirmada exitosamente",
	})
}
